### PR TITLE
COM-139: Textarea component

### PR DESCRIPTION
**Pull request by TheEnmanuel23**
## Please delete checklist items and sections that are not relevant

**Checklist**

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Please check if the PR fulfills these requirements**

- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Storybook or docs have been added / updated (for bug fixes / features)

**What is the ticket / issue associated with this change?** (Link to the ticket or issue)
https://impulsumvc.atlassian.net/browse/COM-139

**Other information**:
- Added engines property to the package.json file in order to get an warning or even an error when using the wrong version of NodeJS.
**Before, when executing `yarn storybook` script and using node v17**

image

**after, executing `yarn storybook` or `yarn install` and using node v17** 
image

- I added the package `@testing-library/jest-dom` to make our tests more declarative and clear to read.

**Storybook Link** (Link to a local storybook url to review new functionality)
https://5eed577398d08d0022b52d76-npvrojhayl.chromatic.com/?path=/story/components-primitives-textarea--all

### CODE REVIEW

File: packages/components/src/components/primitives/Textarea/Textarea.test.tsx [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    expect(textarea).toHaveValue(value);
+  });
+
+  test('should render an invalid textarea', () => {
+    renderWithProviders(<Textarea isInvalid />);
+    const textarea = screen.getByTestId(testId);
+    expect(textarea).toHaveStyle({
+      borderColor: `var(--chakra-colors-alert-error-default)`,
+    });
+  });
```
Comment (tkanzakic) [inline packages/components/src/components/primitives/Textarea/Textarea.test.tsx:28]: no need to unit test styles, we have chromatic for this

File: packages/components/package.json [hunk 2]
```diff
@@ -32,6 +34,7 @@
     "@chakra-ui/theme-tools": "^1.3.1",
     "@emotion/react": "^11",
     "@emotion/styled": "^11",
+    "@testing-library/jest-dom": "^5.16.1",
```
Comment (tkanzakic) [inline packages/components/package.json:37]: do we really dependency?, and in any case, it should be a `devDependency` no a production one

File: packages/components/package.json [hunk 3]
```diff
@@ -45,5 +48,8 @@
     "react": "^17.0.2",
     "react-dom": "^17.0.2",
     "storybook-chromatic": "^4.0.2"
+  },
+  "engines": {
+    "node": ">=10 <=16"
```
Comment (tkanzakic) [inline packages/components/package.json:53]: thanks for adding this!!!, _but_ we require node 16

File: packages/components/src/components/primitives/Textarea/types.ts [hunk 4]
```diff
@@ -0,0 +1,12 @@
+import { TextareaProps as ChakraTextareaProps } from "@chakra-ui/react";
+
+export const textareaSizes = ['s', 'm'] as const;
+export const textareaVariants = ["default"] as const;
```
Comment (tkanzakic) [inline packages/components/src/components/primitives/Textarea/types.ts:4]: this variant name do not match the name of the variant defined in the theme file

**Review by tkanzakic — CHANGES_REQUESTED** [review]
the placeholder color is not the right one in this PR (check how the `Input` component was implemented), also there are some styles that do not match the figma file

File: packages/components/src/components/primitives/Textarea/Textarea.test.tsx (hunk 1 above)
Comment (guilean) [reply packages/components/src/components/primitives/Textarea/Textarea.test.tsx:28]: I think we should be good to remove `import '@testing-library/jest-dom';` then

File: packages/components/src/components/primitives/Textarea/Textarea.tsx [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import React, { FC } from 'react';
+import { Textarea as ChakraTextarea } from '@chakra-ui/react';
+import { TextareaProps } from './types';
+
+export const Textarea: FC<TextareaProps> = (props) => (
+  <ChakraTextarea {...props} data-testid="cmpsr.textarea" />
```
Comment (guilean) [inline packages/components/src/components/primitives/Textarea/Textarea.tsx:6]: Should we reverse the order of `props` and `data-testid`?

File: packages/components/src/components/primitives/Textarea/Textarea.test.tsx (hunk 1 above)
Comment (TheEnmanuel23) [reply packages/components/src/components/primitives/Textarea/Textarea.test.tsx:28]: `@testing-library/jest-dom` can make it easier for us how we validate what is rendered, https://github.com/testing-library/jest-dom#the-problem, but obviously it is one more option

File: packages/components/package.json (hunk 2 above)
Comment (TheEnmanuel23) [reply packages/components/package.json:37]: my bad when adding it in the production dependencies, sorry, but about if it is necessary, well, as they say in their documentation: **These will make your tests more declarative, clear to read and to maintain.**
https://github.com/testing-library/jest-dom#the-problem.
but I can perfectly remove it if you and @guilean don't like it. NP

File: packages/components/package.json [hunk 6]
```diff
@@ -42,10 +42,14 @@
     "@storybook/addon-essentials": "^6.4.13",
     "@storybook/addon-links": "^6.4.13",
     "@storybook/react": "^6.4.13",
+    "@testing-library/jest-dom": "^5.16.1",
```
Comment (tkanzakic) [inline packages/components/package.json:45]: do we need this?

File: packages/components/src/theme/ComposerTheme/Components/Textarea.ts [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+  _placeholder: {
+    color: 'text-secondary',
+  },
+  _hover: {
+    backgroundColor: 'background-action-hover',
+    borderWidth: '0.063rem',
+    borderColor: 'ui-element-outline-active',
+  },
+  _disabled: {
+    backgroundColor: 'background-action-disabled',
```
Comment (tkanzakic) [inline packages/components/src/theme/ComposerTheme/Components/Textarea.ts:17]: missing color for disabled state

File: packages/components/src/theme/ComposerTheme/Components/Textarea.ts [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+  return ({ theme }) => ({
+    ...inputStyles[size],
+    ...theme.textStyles[textStyles[size]],
+  })
+}
+
+export const Textarea: ComponentStyleConfig = {
+  baseStyle,
+  variants: {
+    outline: baseStyle
```
Comment (tkanzakic) [inline packages/components/src/theme/ComposerTheme/Components/Textarea.ts:48]: is this needed?

File: packages/components/src/components/primitives/Textarea/Textarea.stories.tsx [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    size: {
+      options: textareaSizes,
+      control: { type: 'select' },
+    },
+    isInvalid: {
+      control: { type: 'boolean' },
+    },
+    isDisabled: {
+      control: { type: 'boolean' },
+    },
```
Comment (tkanzakic) [inline packages/components/src/components/primitives/Textarea/Textarea.stories.tsx:20]: nit: this is not necessary, storybook detects and handle boolean values

**Review by tkanzakic — CHANGES_REQUESTED** [review]

File: packages/components/src/theme/ComposerTheme/Components/Textarea.ts (hunk 8 above)
Comment (TheEnmanuel23) [reply packages/components/src/theme/ComposerTheme/Components/Textarea.ts:48]: yes, If I remove this property, the styles on focus event don't work:
image

File: packages/components/package.json (hunk 6 above)
Comment (TheEnmanuel23) [reply packages/components/package.json:45]: I am using it to test the textarea component

File: packages/components/src/components/primitives/Textarea/Textarea.stories.tsx (hunk 9 above)
Comment (TheEnmanuel23) [reply packages/components/src/components/primitives/Textarea/Textarea.stories.tsx:20]: done!

**Review by tkanzakic — CHANGES_REQUESTED** [review]
The disabled state do not match figma, see attached image (left: storybook; right: figma)

Screenshot 2022-02-01 at 10 39 03

**Comment by TheEnmanuel23** [comment]
> The disabled state do not match figma, see attached image (left: storybook; right: figma)
> 
> Screenshot 2022-02-01 at 10 39 03

**Review by tkanzakic — CHANGES_REQUESTED** [review]
the border radius in figma is of `0.375rem` (6px) and the component have `0.4rem`

**Review by guilean — COMMENTED** [review]
PR has some conflicts. Is ready to be re-reviewed? @TheEnmanuel23

**Review by TheEnmanuel23 — COMMENTED** [review]
@guilean, it's ready for review.
I was chatting wit Rik about the padding in Large sizes since the design shows a wrong value, he said the correct value is 16px  on the side paddings for both flushed and outline variants

**Review by tkanzakic — APPROVED** [review]

**Pull request merged**